### PR TITLE
Fix missing outputs for jobs with matrix

### LINKED ISSUE #32795: Bug with job outputs when using matrix jobs
### Description

Hello, I found a bug which affected outputs when using matrix jobs.

Here an example of jobs: https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#using-job-outputs-in-a-matrix-job

```
jobs:
  job1:
    runs-on: ubuntu-latest
    outputs:
      output_1: ${{ steps.gen_output.outputs.output_1 }}
      output_2: ${{ steps.gen_output.outputs.output_2 }}
      output_3: ${{ steps.gen_output.outputs.output_3 }}
    strategy:
      matrix:
        version: [1, 2, 3]
    steps:
      - name: Generate output
        id: gen_output
        run: |
          version="${{ matrix.version }}"
          echo "output_${version}=${version}" >> "$GITHUB_OUTPUT"
  job2:
    runs-on: ubuntu-latest
    needs: [job1]
    steps:
      # Will show
      # {
      #   "output_1": "1",
      #   "output_2": "2",
      #   "output_3": "3"
      # }
      - run: echo '${{ toJSON(needs.job1.outputs) }}'
```

However it shows in Gitea:
```
{
  "output_1": "1",
  "output_2": "",
  "output_3": ""
}
```

I think the issue in evaluating outputs for every matrix job:
```
🏁  Job succeeded
expression '${{ steps.gen_output.outputs.output_1 }}' rewritten to 'format('{0}', steps.gen_output.outputs.output_1)'
evaluating expression 'format('{0}', steps.gen_output.outputs.output_1)'
expression 'format('{0}', steps.gen_output.outputs.output_1)' evaluated to '%!t(string=)'
expression '${{ steps.gen_output.outputs.output_2 }}' rewritten to 'format('{0}', steps.gen_output.outputs.output_2)'
evaluating expression 'format('{0}', steps.gen_output.outputs.output_2)'
expression 'format('{0}', steps.gen_output.outputs.output_2)' evaluated to '%!t(string=)'
expression '${{ steps.gen_output.outputs.output_3 }}' rewritten to 'format('{0}', steps.gen_output.outputs.output_3)'
evaluating expression 'format('{0}', steps.gen_output.outputs.output_3)'
expression 'format('{0}', steps.gen_output.outputs.output_3)' evaluated to '%!t(string=3)'
```

Could you please fix it and provide a workaround for now?

Thanks!

### Gitea Version

v1.22.4

### PR DESCRIPTION
**Pull request by Zettat123**
Fix #32795

If a job uses a matrix, multiple `ActionRunJobs` may have the same `JobID`. We need to merge the outputs of these jobs to make them available to the jobs that need them.

### CODE REVIEW

**Comment by lunny** [comment]
Can you add some tests?

**Comment by Zettat123** [comment]
> Can you add some tests?

I'll try to add some tests

File: routers/api/actions/runner/utils.go [hunk 1]
```diff
@@ -162,28 +162,56 @@ func findTaskNeeds(ctx context.Context, task *actions_model.ActionTask) (map[str
...
+func mergeTwoOutputs(o1, o2 map[string]string) map[string]string {
+	ret := make(map[string]string, len(o1))
+	for k1, v1 := range o1 {
+		if len(v1) > 0 {
+			ret[k1] = v1
+		} else {
+			ret[k1] = o2[k1]
+		}
+	}
+	return ret
```
Comment (wxiaoguang) [inline routers/api/actions/runner/utils.go:216]: Does it need to copy the renaming keys (not in o1) from o2?

Comment (Zettat123) [reply routers/api/actions/runner/utils.go:216]: I think there should be no "remaining keys". Take the following job(from [GH doc](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#using-job-outputs-in-a-matrix-job)) as an example

```
  job1:
    runs-on: ubuntu-latest
    outputs:
      output_1: ${{ steps.gen_output.outputs.output_1 }}
      output_2: ${{ steps.gen_output.outputs.output_2 }}
      output_3: ${{ steps.gen_output.outputs.output_3 }}
    strategy:
      matrix:
        version: [1, 2, 3]
    steps:
      - name: Generate output
        id: gen_output
        run: |
          version="${{ matrix.version }}"
          echo "output_${version}=${version}" >> "$GITHUB_OUTPUT"
```

The output names are defined in `job1.outputs`. Although there will be 3 jobs(for version 1, 2, 3), their output names will be the same, i.e. every job will output `output_1`, `output_2` and `output_3`, even though some of their values ​​may be empty.

**Comment by Zettat123** [comment]
> Can you add some tests?

It is not easy to test because the function is part of the process of fetching tasks by act_runner. Right now we don't have a mechanism to test this. I would like to add a mechanism to test such functions after finishing the PRs I'm working on.

**Comment by wxiaoguang** [comment]
I think `findTaskNeeds` could be tested by some mock data? It only does some database operations?

**Comment by Zettat123** [comment]
> I think `findTaskNeeds` could be tested by some mock data? It only does some database operations?

Ah, yes, Only testing `findTaskNeeds` is possible. I think I can add a test for `findTaskNeeds` for now and add integration tests for fetching tasks later.

**Review by wxiaoguang — APPROVED** [review]
Vote my LGTM ahead

**Review by lunny — APPROVED** [review]

**Pull request merged**